src/loaders/lottie/tvgLottieBuilder.cpp

```diff
@@ -1071,21 +1094,43 @@ void LottieBuilder::updateText(LottieLayer* layer, float frameNo)
                     else if ((*s)->based == LottieTextRange::Based::Lines) basedIdx = line;
 
                     if (basedIdx < start || basedIdx >= end) continue;
+
                     auto& matrix = shape->transform();
 
+                    // TextGroup transformation is performed once
+                    if (textGroup->paints().size() == 0) {
+                        auto alignment = text->alignmentOption.groupAlignment(frameNo);
+
+                        // center pivoting
+                        textGroupMatrix.e13 += alignment.x;
+                        textGroupMatrix.e23 += alignment.y;
+
+                        rotate(&textGroupMatrix, (*s)->style.rotation(frameNo));
+
+                        auto pivotX = alignment.x * -1;
+                        auto pivotY = alignment.y * -1;
+
+                        //center pivoting
+                        textGroupMatrix.e13 += (pivotX * textGroupMatrix.e11 + pivotX * textGroupMatrix.e12);
+                        textGroupMatrix.e23 += (pivotY * textGroupMatrix.e21 + pivotY * textGroupMatrix.e22);
+                        
+                        textGroup->transform(textGroupMatrix);
+                    }
+
                     shape->opacity((*s)->style.opacity(frameNo));
 
                     auto color = (*s)->style.fillColor(frameNo);
                     shape->fill(color.rgb[0], color.rgb[1], color.rgb[2], (*s)->style.fillOpacity(frameNo));
 
-                    rotate(&matrix, (*s)->style.rotation(frameNo));
-
                     auto glyphScale = (*s)->style.scale(frameNo) * 0.01f;
                     tvg::scale(&matrix, glyphScale.x, glyphScale.y);
 
                     auto position = (*s)->style.position(frameNo);
                     translate(&matrix, position.x, position.y);
 
+                    // NOTE: duplicated with Mira's PR, should be removed
+                    shape->transform(matrix);
```

Comment:
Please ignore this changes, it's workaround to fix current rotation bug. Which will be closed in #2891 patch. After all, I'll remove this commit and line.

Alternatively, you can review after I rebase.

Reply:
@tinyjin could you please rebase PR?